### ISSUE TITLE
Fixed chunk offsets (feature)

### ISSUE THREAD

**Issue by jamespharaoh** [zbackup/zbackup#133]
I propose to add an option to zbackup which makes it either always deduplicate on multiples of a specified size, or perhaps to prefer it. I think this might improve backup efficiency when the input is known to be aligned in a certain way, eg a tar file where files and metadata blocks are always on 512 byte boundaries.

It should also be possible to attempt a backup based on a smaller chunk size. Ideally we would keep a much bigger index, including ranges of chunks, and this relates to my previous issue #131.

I understand that the current argument against a smaller chunk size, or index thereof, is that the efficiency of the lookup during decompression becomes more expensive. However, this cost is /drastically/ reduced if we only look for duplication on specific offsets.

Potentially, this could be implemented as an "extra" level of indexing, which is only invoked on a multiple of the specified size boundary.

In my main use case, which involves tar files, this would be very likely to improve the efficiency in terms of storage and speed. I have no plans to implement anything at the moment, but would like to if there is community support.

Comments please?

**Comment by mknjc** [comment]
This would need a test implementation to check if the benefit is big enough.
Index lookup could be made a lot better with a tree structure or a bitmap...

**Comment by jamespharaoh** [comment]
Yes I realise there's lots of work to do. I think the first step, however, is to agree what this would look like on-disk, and how the compatibility would be managed. I will go ahead and create a branch with a proposal for this, and a basic implementation and test.

**Comment by mknjc** [comment]
What is your plan to get the deduplication right? You have to store the rolling hash and a checksum for your 512 byte chunk somewhere, (in the index?) so 8 bytes + 32 bytes = 40 bytes per 512 byte data.

**Comment by jamespharaoh** [comment]
Well as I've explained, I'm more interested in this in the long term, but would very much prefer there to be an agreed on-disk format for storing the partial chunk references, and for this to be supported by native zbackup. I would be wary of recommending anyone use it until that is well established.

In terms of how I do the deduplication at this level, there are loads of ways I can cheat. For example, I may simply do deduplication based on identical filenames, and use a very strong hash to further verify them. In this case, I'd be implementing the backup process myself, but that seems reasonable for a first attempt at this.

Your comments about a tree structure make sense, I haven't thought in detail about how this would work but probably would look for some kind of preexisting database implementation which already does this kind of optimisation for me. I believe most database indexes use prefix optimisation, which I think you are referring to.
